src/Microsoft.OData.Core/ObjectPool.cs

```diff
@@ -0,0 +1,100 @@
+﻿//---------------------------------------------------------------------
+// <copyright file="ObjectPool.cs" company="Microsoft">
+//      Copyright (C) Microsoft Corporation. All rights reserved. See License.txt in the project root for license information.
+// </copyright>
+//---------------------------------------------------------------------
+
+using System;
+
+namespace Microsoft.OData
+{
+    /// <summary>
+    /// Manages a pool of reusable objects. An object retrieved from the pool should be returned when it's no longer being used so that it can be later recycled.
+    /// </summary>
+    /// <typeparam name="T">The type of objects to pool.</typeparam>
+    /// <remarks>This object pool is not thread-safe and not intended to be shared across concurrent requests.
+    /// Consider using the DefaultObjectPool in 8.x once we drop support for netstandard 1.1
+    /// </remarks>
+    internal class ObjectPool<T> where T : class

```

Comment:
I might be trivializing this, but why isn't this the implementation:
```
public sealed class ObjectPool<T>
{
  private readonly Func<T> factory;
  private readonly T[] items;
  private readonly int next;
  private readonly numberOfObjectsCreated;

  public ObjectPool(Func<T> factory, int size)
  {
    this.factory = factory;
    this.items = new T[size];
    this.next = 0;
    this.numberOfObjectsCreated = 0;
  }

  public T Get()
  {
    // do we need to create new items?
    if (this.next == this.numberOfObjectsCreated)
    {
      // are we out of slots for new items?
      if (this.numberOfObjectsCreated > this.items.Length)
      {
        throw new Exception("no more available");
      }

      this.items[this.next] = this.factory();
      this.numberOfObjectsCreated++;
    }

    return this.items[this.next++];
  }

  public void Return()
  {
    this.next--;
  }
}
```

Reply:
I'm not sure if it's a good idea to throw an exception when we're out of slots. I think in that case we should either "resize" the array (which is maybe more costly than we want to pay) or create a new object that's not stored in the pool. If there's more demand than what can fit in the pool, then the excess is not pooled, but doesn't throw either. My 2 cents.

Reply:
I've consolidated my different feedback threads into this one because I feel like something is getting lost. 

1. The majority of my feedback revolves around the fact that we are not just using indices to track the next items. Doing so removes so much of the code, and seems to make the total number of operations consistent for each scenario. Particularly, notice how the implementation above is O(1), while the current implementation is O(n). There is a lot of extra code to try to make it mostly perform as O(1), but I don't think we should be optimizing an O(n) algorithm when an O(1) option is available. 
2. By being fixed size, we lose out on the optimization if the nesting level ever changes. I'm not actually proposing that we just throw an exception if the nesting level changes, and that the customer ends up perplexed. I'm saying that a fixed size `ObjectPool` should have in its contract that it throws when it runs out of space. We should catch this exception somewhere else (probably in the `ODataSerializer`). If we don't throw an exception and we increase the nesting level, we will never notice that we are losing out on this optimization. I think we should either use a dynamically sized object pool, or we should throw an exception that we surface in `Debug` configurations and don't surface in `Release` configurations. 

Ultimately, I think we should be using the .NET object pool implementation and we don't use an object pool for versions that don't have it. This should incentivize users to move to newer versions of .NET in order to receive these performance benefits.

Reply:
@corranrogue9 
1. Your solution still doesn't the resolve the issue where we have a fixed size object pool. We shouldn't be throwing an exception.
2. My solution is flexible when the nesting is larger than the array size. We will still spin more objects. What will happen is when returning the objects to the pool, the extra objects will be disposed. The solution is modelled around the .Net object pool.
3. In the ObjectPool<T> docstring I have indicated we should switch to the DefaultObjectPool in 8.x once we drop support for `netstandard 1.1`. Having 2 types object pools in ODL now only adds extra complexity.

Conclusion: This PR was intended on to reduce heap allocation. At the same time not regress cpu time. Further optimizations can be done later.

Reply:
> Your solution still doesn't the resolve the issue where we have a fixed size object pool

It does. It throws an exception.

> We shouldn't be throwing an exception

Yes we should. We need to notice when we have made an unrelated change that causes us to go beyond the current nesting level and therefore lose out on this optimization. *Please address this concern that I have*: if we don't throw an exception, how do we, as a team, know that we have inadvertently increased the nesting level?

> My solution is flexible when the nesting is larger than the array size. We will still spin more objects. 

What is the benefit of doing this? From what I see, it only serves to hide potential bugs. 

> The solution is modelled around the .Net object pool.

The .NET implementation has an extra requirement of being thread-safe, meaning that `Interlocked.Exchange` is *imperative* to its functioning. Because of this, reference types are needed and keeping a reference to the next item is essential. Without that requirement, we can improve performance by using a different algorithm rather than by optimizing the .NET algorithm.

> In the ObjectPool docstring I have indicated we should switch to the DefaultObjectPool in 8.x once we drop support for netstandard 1.1

Please create an issue for this so that the work can be planned for and assigned

> Having 2 types object pools in ODL now only adds extra complexity

I am not proposing that we have 2. I am proposing that we don't ship this optimization for older versions of .NET. This will incentivize customers to use newer versions of .NET. Alternatively, we could simply take the .NET implementation and place it directly into our repo, referencing where it came from. Then, we can ship down-level, but have complete consistency across versions.

Reply:
I don't think it's a good idea to throw an exception, because that's implying that having responses with depths that exceed the object pool size is an error. The size of the object pool shouldn't impose a limit on the depth of the response (there's a separate setting for that).

If the goal of the exception is for sanity checks to ensure we are not exceeding the object pool size due a bug in our code rather than due to the depth of the response, maybe we could have debug messages or assertions when the object pool size is exceeded, when the number of returned objects is not equal the number of requested objects, etc.

Or did you mean that we conditionally throw exceptions based on `#if DEBUG` or something like that?

I prefer the simpler implementation provided by @corranrogue9, but I still think it's valuable to delay allocating the array until the second element is requested, that way we avoid unnecessary allocations for non-nested responses. Unless we can demonstrate that the cost of having the `if` statement that's always checked is worse.

Also, since the point of this PR is to reduce allocations (and GC), I think it's okay to only care about reference types.